### ISSUE TITLE
Increases complexity and doesn't really solve the problem. 

### ISSUE THREAD

**Issue by ffigiel** [piecioshka/test-es2015-proxy#3]
With proxies you can throw errors in case of a missing property, but that solution raises another problem: you can't really use plain objects anymore. You'd want all objects to raise missing property exceptions. If some objects are plain and some are behind your proxy, you will lose control over your code, not remembering which object is safe and which is not. 

To handle current behavior with missing properties, you can do three things:
- Writing extensive integration tests. (I don't think this is a maintainable solution)
- Use Flow to add type checks to your code
- Use a different language like TypeScript or Elm

**Comment by piecioshka** [comment]
@megapctr hi there!
1. Can you provide me any example of another problem what this code raise?
2. What does it means: `Writing extensive integration tests`?
3. Can you provide me example of solutions in Flow & TypeScript.

Thanks.

**Comment by ffigiel** [comment]
1) the problem lies within discipline. Every developer has to pass every object they receive through your proxy. If they forget, some parts of application will still use plain objects, and the problem with undefined returns. Besides, it's not very convenient to pass every object through some function before you can use it. It also decreases readability and maintainability of the code.

2) writing tests means writing a program that will use your application and check if all results are valid. For example, a `calculator` function given a sequence `'2+2='`  should return `4`. Following this example, an integration test would test something that _uses_ the calculator function in some way, e.g. through an user interface with a text input.

My point about testing is that it's impossible to test your application in every possible way, because that is too expensive compared to other solutions (e.g. Flow). That doesn't mean you shouldn't write any tests (you should).  

3) I didn't use flow and ts much, but the point is that your function can expect an object with a specific shape. For example, it may expect a user object that has a number field 'age' and a string field 'name'.

**Comment by ffigiel** [comment]
For examples on flow, you can check their [documentation on objects](https://flowtype.org/docs/objects.html)

**Comment by piecioshka** [comment]
@ffigiel you're right, and thanks for the detailed write-up. This repo was only ever an experiment to see what ES2015 `Proxy` makes possible, not a pattern to actually use.

Your points hold up: the approach relies on discipline (every object has to be wrapped, or that part of the app silently goes back to `undefined`), it makes it hard to tell which objects are "safe", it only fails at runtime on the executed path, and the trap adds per-access overhead. Static typing with TypeScript or Flow catches missing/misspelled properties at compile time with zero runtime cost and nothing to wrap, which is clearly the better tool. Integration tests help but are far more expensive for the same guarantee.

I've captured this in a new "Limitations" section in the README (#8) so the repo is honest about being a demo rather than advice. Closing as answered.
